### ISSUE TITLE
ttnn.batch_norm causing PCC Drop with program cache enabled

### ISSUE THREAD

**Issue by VirdhatchaniKN** [tenstorrent/tt-metal#20877]
### Description of the bug :
While enabling program cache for the Pytorch project, an issue with low PCC for a certain op sequence was found. In particular, the PCC is low for `ttnn.batch_norm` followed by `ttnn.floor_div`. Repro steps are below.

### Steps to reproduce the issue : 
[Check here](https://github.com/tenstorrent/tt-metal/issues/20877#issuecomment-2818719649)

### Expected Behavior : 

The program cache should only return a hit when both input shapes and values are compatible. If input values change in a way that affects computation, a cache miss should occur to ensure correctness.

**Comment by VirdhatchaniKN** [comment]
@jmalone-tt , 
Please add the necessary information regarding the failure

**Comment by jmalone-tt** [comment]
Took a bit to reduce this to a minimal example, thanks for your patience! The following code reliably reproduces a PCC failure for me (some boiler plate for PCC comparison, actual test is in the lower portion):
```python
import torch
import ttnn

import numpy as np

def comp_pcc(golden, calculated, pcc=0.99):
    golden = torch.Tensor(golden)
    calculated = torch.Tensor(calculated)

    if golden.dtype != calculated.dtype:
        calculated = calculated.type(golden.dtype)

    if torch.all(torch.isnan(golden)) and torch.all(torch.isnan(calculated)):
        # logger.warning("Both tensors are 'nan'")
        return True, 1.0

    if torch.all(torch.isnan(golden)) or torch.all(torch.isnan(calculated)):
        # logger.error("One tensor is all nan, the other is not.")
        return False, 0.0

    # Test if either is completely zero
    if torch.any(golden.bool()) != torch.any(calculated.bool()):
        # logger.error("One tensor is all zero")
        return False, 0.0

    # For now, mask all infs and nans so that we check the rest... TODO
    golden = golden.clone()
    golden[
        torch.logical_or(
            torch.isnan(golden),
            torch.logical_or(torch.isinf(golden), torch.isneginf(golden)),
        )
    ] = 0
    calculated = calculated.clone()
    calculated[
        torch.logical_or(
            torch.isnan(calculated),
            torch.logical_or(torch.isinf(calculated), torch.isneginf(calculated)),
        )
    ] = 0

    if torch.equal(golden, calculated):
        return True, 1.0

    if golden.dtype == torch.bfloat16:
        golden = golden.type(torch.float32)
        calculated = calculated.type(torch.float32)
    cal_pcc = np.min(
        np.ma.corrcoef(
            np.ma.masked_invalid(torch.squeeze(golden).detach().numpy()).flatten(),
            np.ma.masked_invalid(torch.squeeze(calculated).detach().numpy()).flatten(),
        )
    )

    if isinstance(cal_pcc, np.ma.core.MaskedConstant):
        return True, 1.0

    return cal_pcc >= pcc, cal_pcc

def assert_with_pcc(expected_pytorch_result, actual_pytorch_result, pcc=0.999):
    assert list(expected_pytorch_result.shape) == list(
        actual_pytorch_result.shape
    ), f"list(expected_pytorch_result.shape)={list(expected_pytorch_result.shape)} vs list(actual_pytorch_result.shape)={list(actual_pytorch_result.shape)}"
    pcc_passed, pcc_message = comp_pcc(expected_pytorch_result, actual_pytorch_result, pcc)
    assert pcc_passed, f"pcc check failed (expected {pcc} but got {pcc_message})"
    return pcc_passed, pcc_message

def floor_divide_forward(device, arg0_1, arg1_1):
    ttnn_from_torch = ttnn.from_torch(arg0_1, device = device, layout = ttnn.TILE_LAYOUT, dtype = ttnn.bfloat16)
    ttnn_from_torch_1 = ttnn.from_torch(arg1_1, device = device, layout = ttnn.TILE_LAYOUT, dtype = ttnn.bfloat16)
    ttnn_floor_div = ttnn.floor_div(ttnn_from_torch, ttnn_from_torch_1)
    ttnn_to_torch = ttnn.to_torch(ttnn_floor_div, dtype = torch.bfloat16)
    return ttnn_to_torch


def batch_norm_inference_forward(device, arg0_1, arg1_1, arg2_1, arg3_1, arg4_1):
    ttnn_from_torch = ttnn.from_torch(arg2_1, device = device, layout = ttnn.TILE_LAYOUT, dtype = ttnn.bfloat16)
    ttnn_add = ttnn.add(ttnn_from_torch, 1e-05)
    ttnn_rsqrt = ttnn.rsqrt(ttnn_add)
    ttnn_reshape = ttnn.reshape(ttnn_rsqrt, (1, 100, 1, 1))
    ttnn_from_torch_1 = ttnn.from_torch(arg1_1, device = device, layout = ttnn.TILE_LAYOUT, dtype = ttnn.bfloat16)
    ttnn_reshape_1 = ttnn.reshape(ttnn_from_torch_1, (1, 100, 1, 1))
    ttnn_from_torch_2 = ttnn.from_torch(arg0_1, device = device, layout = ttnn.TILE_LAYOUT, dtype = ttnn.bfloat16)
    ttnn_subtract = ttnn.subtract(ttnn_from_torch_2, ttnn_reshape_1)
    ttnn_multiply = ttnn.multiply(ttnn_subtract, ttnn_reshape)
    ttnn_from_torch_3 = ttnn.from_torch(arg3_1, device = device, layout = ttnn.TILE_LAYOUT, dtype = ttnn.bfloat16)
    ttnn_reshape_2 = ttnn.reshape(ttnn_from_torch_3, (1, 100, 1, 1))
    ttnn_multiply_1 = ttnn.multiply(ttnn_multiply, ttnn_reshape_2)
    ttnn_from_torch_4 = ttnn.from_torch(arg4_1, device = device, layout = ttnn.TILE_LAYOUT, dtype = ttnn.bfloat16)
    ttnn_reshape_3 = ttnn.reshape(ttnn_from_torch_4, (1, 100, 1, 1))
    ttnn_add_1 = ttnn.add(ttnn_multiply_1, ttnn_reshape_3)
    ttnn_to_torch = ttnn.to_torch(ttnn_add_1, dtype = torch.bfloat16)
    return ttnn_to_torch

if __name__ == "__main__":
    # open device
    device = ttnn.open_device(device_id=0, l1_small_size=16384)
    ttnn.enable_program_cache(device)

    # run floor_divide
    numerator = torch.randint(1, 15, (1, 12, 9, 9)).to(torch.bfloat16)
    denominator = torch.randint(1, 15, ()).to(torch.bfloat16)
    floor_divide_forward(device, numerator, denominator)

    # run batch_norm_inference
    input_tensor = torch.randn((1, 100, 14, 14), dtype=torch.bfloat16)
    mean = torch.randn(100, dtype=torch.bfloat16)
    var = torch.randn(100, dtype=torch.bfloat16) ** 2
    weight = torch.randn(100, dtype=torch.bfloat16)
    bias = torch.randn(100, dtype=torch.bfloat16)
    args = input_tensor, mean, var, weight, bias

    # run
    result_before = torch.nn.functional.batch_norm(*args)
    ttnn_result = batch_norm_inference_forward(device, *args)

    assert_with_pcc(result_before, ttnn_result, pcc=0.99)

    # close device
    ttnn.close_device(device)
```

**Comment by cmaryanTT** [comment]
When root caused, please broadcast the details on the eltwise slack channel, for learning purposes.

**Comment by VirdhatchaniKN** [comment]
Merged the custom hash function for BN to main.
The issue was caused in the custom batch norm function as in above test file,due to ttnn.multiply and mentioned in #21086 and not in ttnn.batch_norm.
